<a href="https://colab.research.google.com/github/EduardaDalmas/python_bigO/blob/main/Teste_comparativo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Trabalho Prático (10pt)


1. Agenda de Compromissos:

Crie um vetor ordenado para armazenar os compromissos, onde cada compromisso é um objeto com informações como data, hora, descrição, etc. 

a) Implemente um algoritmo para adicionar e excluir compromissos no vetor ordenado, mantendo sempre a ordem cronológica dos compromissos.

b) Implemente um algoritmo de busca binária para permitir aos usuários procurarem por compromissos específicos por data, hora ou outra informação relevante.

c) Implemente também a visualização de todos os compromissos em ordem cronológica.

# Código de Compromissos

In [4]:
import pprint

In [5]:
import random

In [6]:
import datetime

In [7]:
import numpy as np

In [476]:
class Compromisso:
  def __init__(self, data, hora, descricao):
    self.data = data
    self.hora = hora
    self.descricao = descricao

  def __lt__(self, other):
     return self.data < other.data

  def __str__(self):
    return f"{self.data} - {self.hora} - {self.descricao}"


class Agenda():
  def __init__(self, capacidade):
    self.capacidade = capacidade
    self.ultima_posicao = -1
    self.valores = [None] * self.capacidade

  def adicionar(self, valor):
    if self.ultima_posicao == self.capacidade -1:
      print('Capacidade máxima atingida')
      return 
    posicao = 0
    for i in range(self.ultima_posicao + 1):
      posicao = i
      if self.valores[i] > valor:
        break
      if i == self.ultima_posicao:
        posicao = i + 1
    x = self.ultima_posicao
    while x >= posicao:
      self.valores[x + 1] = self.valores[x]
      x -= 1
    self.valores[posicao] = valor
    self.ultima_posicao += 1 

  def excluir(self, valor):
    if self.ultima_posicao == -1:
      print('Array vazio')
      return 
    else:
      for i in range(valor, self.ultima_posicao):
        self.valores[i] = self.valores[i + 1]
      self.ultima_posicao -= 1

  def buscar(self, valor, tipo):
    limite_inferior = 0
    limite_superior = self.ultima_posicao

    while True:
      posicao_atual = int((limite_inferior + limite_superior) / 2)

      if tipo == 'objeto':
        if self.valores[posicao_atual] == valor:
          return posicao_atual

        elif limite_inferior > limite_superior:
          return -1
        
        else:
          if self.valores[posicao_atual] < valor:
            limite_inferior = posicao_atual + 1
          else:
            limite_superior = posicao_atual - 1
      
      if tipo == 'data':
        if self.valores[posicao_atual].data == valor:
          return posicao_atual

        elif limite_inferior > limite_superior:
          return -1
        
        else:
          if self.valores[posicao_atual].data < valor:
            limite_inferior = posicao_atual + 1
          else:
            limite_superior = posicao_atual - 1
      
      if tipo == 'hora':
        if self.valores[posicao_atual].hora == valor:
          return posicao_atual

        elif limite_inferior > limite_superior:
          return -1
        
        else:
          if self.valores[posicao_atual].hora < valor:
            limite_inferior = posicao_atual + 1
          else:
            limite_superior = posicao_atual - 1

      if tipo == 'descricao':
        if self.valores[posicao_atual].descricao == valor:
          return posicao_atual

        elif limite_inferior > limite_superior:
          return -1
        
        else:
          if self.valores[posicao_atual].descricao < valor:
            limite_inferior = posicao_atual + 1
          else:
            limite_superior = posicao_atual - 1      

  def visualizar(self):
    if self.ultima_posicao == -1:
      print('O vetor está vazio')   
    else:
      for i in range(self.ultima_posicao +1):
        compromisso = self.valores[i]
        print('posicao: ', i, ' | data: ', compromisso.data, ' | hora: ', compromisso.hora, ' | descricao: ', compromisso.descricao) 


*Inserir e listar*

In [477]:
agendar = Agenda(1000)

In [478]:
agendar.visualizar()

O vetor está vazio


*Adicionando registros*

In [411]:
for i in range(1000):
  nomes = ['Reunião', 'Jantar', 'Call', 'Retrospectiva', 'Avaliação do time']
  tipo = ['de negocios', 'no Teams', 'no Meet', 'no Discord',]
  descricao_compromisso = random.choice(nomes) + " " + random.choice(tipo)
  ano = random.randint(1900, 2023)
  mes = random.randint(1, 12)
  if mes == 2:  # Fevereiro tem 28 ou 29 dias
    if ano % 4 == 0 and (ano % 100 != 0 or ano % 400 == 0):
        dia = random.randint(1, 29)
    else:
        dia = random.randint(1, 28)
  elif mes in (4, 6, 9, 11):  # Abril, Junho, Setembro e Novembro tem 30 dias
    dia = random.randint(1, 30)
  else:  # Os outros meses tem 31 dias
    dia = random.randint(1, 31)
  data_randomica = datetime.datetime(ano, mes, dia).strftime("%d/%m/%Y")
  hora = datetime.time(random.randint(0, 23), random.randint(0, 59)).strftime("%H:%M")
  novo_agendamento = Compromisso( 
    data_randomica,
    hora,
    descricao_compromisso)
  agendar.adicionar(novo_agendamento)

*Calculando tempo de inserção de 1000 registros*

In [479]:
elementos = []

for i in range(1000):
  nomes = ['Reunião', 'Jantar', 'Call', 'Retrospectiva', 'Avaliação do time']
  tipo = ['de negocios', 'no Teams', 'no Meet', 'no Discord',]
  descricao_compromisso = random.choice(nomes) + " " + random.choice(tipo)
  ano = random.randint(1900, 2023)
  mes = random.randint(1, 12)
  if mes == 2:  # Fevereiro tem 28 ou 29 dias
    if ano % 4 == 0 and (ano % 100 != 0 or ano % 400 == 0):
        dia = random.randint(1, 29)
    else:
        dia = random.randint(1, 28)
  elif mes in (4, 6, 9, 11):  # Abril, Junho, Setembro e Novembro tem 30 dias
    dia = random.randint(1, 30)
  else:  # Os outros meses tem 31 dias
    dia = random.randint(1, 31)
  data_randomica = datetime.datetime(ano, mes, dia).strftime("%d/%m/%Y")
  hora = datetime.time(random.randint(0, 23), random.randint(0, 59)).strftime("%H:%M")
  novo_agendamento = Compromisso( 
    data_randomica,
    hora,
    descricao_compromisso)
  elementos.append(novo_agendamento)

In [480]:
def insercao_agenda(elementos):
  for i in elementos:
    agendar.adicionar(i)
  return agendar

In [433]:
insercao_agenda(elementos)

In [481]:
%timeit insercao_agenda(elementos)

A saída de streaming foi truncada nas últimas 5000 linhas.
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima ating

*Visualizando os registros inseridos*

In [ ]:
agendar.visualizar()

*Excluir*

In [354]:
agendar.excluir(1)

*Buscar*

In [349]:
agendar.buscar(valor='Reunião no Teams', tipo='descricao')

937

In [404]:
agendar.buscar(valor='09/12/1924', tipo='data')

-1

*Busca por 1000 registros*

In [455]:
def busca_agenda(elementos):
  for i in elementos:
    agendar.buscar(valor=i.descricao, tipo='descricao')
  return agendar


In [445]:
busca_agenda(elementos)

In [456]:
%timeit busca_agenda(elementos)

6.95 ms ± 2.19 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


2. Estoque de uma Loja Online:

Crie um catálogo de produtos em um vetor não ordenado, onde cada elemento contém informações sobre o produto, como nome, código, quantidade, preço e avaliação de clientes. 

a) Implemente um algoritmo para adicionar e excluir produtos no vetor não ordenado.

b) Implemente  um algoritmo de busca para permitir aos funcionários procurarem por produtos específicos por código, nome ou outra informação relevante.

c) Implemente um algoritmo de busca ou ordenação para encontrar os melhores produtos, por exemplo, aqueles com as melhores avaliações ou melhores preços. Retorne os produtos encontrados para exibição aos usuários.

d) Implemente também a visualização de todos os produtos do estoque e cálculo de total de estoque.

In [482]:
class Produto:
  def __init__(self, nome, codigo, quantidade, preco, avaliacao):
    self.nome = nome
    self.codigo = codigo
    self.quantidade = quantidade
    self.preco = preco
    self.avaliacao = avaliacao

class Estoque():
  def __init__(self, capacidade):
    self.capacidade = capacidade
    self.ultima_posicao = -1
    self.produtos = [None] * self.capacidade

  def adicionar(self, valor):
    if self.ultima_posicao == self.capacidade -1:
      print('Capacidade máxima atingida')
    else:
      self.ultima_posicao  += 1
      self.produtos[self.ultima_posicao] = valor
  
  def buscar(self, valor, tipo):
    if tipo == 'codigo':
      for i in range(self.ultima_posicao + 1):
        codigo_produto = self.produtos[i]
        if int(valor) == int(codigo_produto.codigo):
          return i
      return -1

    if tipo == 'nome':
      for i in range(self.ultima_posicao + 1):
        codigo_produto = self.produtos[i]
        if valor == codigo_produto.nome:
          return i
      return -1
  
  def ordenar(self, tipo):
    if tipo == 'preco':
      produtos_nao_vazios = filter(lambda produto: produto is not None, self.produtos)
      ordenacao = sorted(produtos_nao_vazios, key=lambda produto: produto.preco, reverse=True)
      for i, produto in enumerate(ordenacao):
        print(f'indice {i} - Nome: {produto.nome} - código: {produto.codigo}, quantidade: {produto.quantidade}, preço: {produto.preco}, avaliações: {produto.avaliacao}')
    
    if tipo == 'avaliacao':
      produtos_nao_vazios = filter(lambda produto: produto is not None, self.produtos)
      ordenacao = sorted(produtos_nao_vazios, key=lambda produto: produto.avaliacao, reverse=True)
      for i, produto in enumerate(ordenacao):
        print(f'indice {i} - Nome: {produto.nome} - código: {produto.codigo}, quantidade: {produto.quantidade}, preço: {produto.preco}, avaliações: {produto.avaliacao}')

  def excluir(self, valor, tipo):
    posicao = self.buscar(valor, tipo)
    if posicao == -1:
       return -1
    else:
      for i in range(posicao, self.ultima_posicao):
        self.produtos[i] = self.produtos[i + 1]
      self.ultima_posicao -= 1

  def visualizar(self):
    if self.ultima_posicao == -1:
      print('O vetor está vazio')   
    else:
      for i in range(self.ultima_posicao +1):
        produto = self.produtos[i]
        print('posicao: ', i, ' | nome: ', produto.nome, ' | codigo: ', produto.codigo, ' | quantidade: ', produto.quantidade, ' | preco: ', produto.preco, ' | avaliacao: ', produto.avaliacao) 
        
  def calcular(self): 
    for i in range(self.ultima_posicao +1):
      produto = self.produtos[i]
      total = int(produto.quantidade) * int(produto.preco)
    return total

*Insere e lista*

In [485]:
estocar = Estoque(1000)

In [484]:
for i in range(10):
  nomes = ['Livro', 'Caderno', 'Caneta', 'Borracha', 'Lapis']
  marcas = ['Big', 'Fabber Castel', 'Foroni', 'Disney', 'Marvel']
  nome_produto = random.choice(nomes) + " " + random.choice(marcas)
  novo_produto = Produto(
    nome_produto,
    str(random.randint(1, 1000)),
    str(random.randint(1, 100)),
    str(random.randint(1, 10)),
    str(random.randint(1, 5)))
  estocar.adicionar(novo_produto)

In [489]:
estocar.visualizar()

posicao:  0  | nome:  Caderno Fabber Castel  | codigo:  527  | quantidade:  30  | preco:  1  | avaliacao:  3
posicao:  1  | nome:  Caneta Foroni  | codigo:  34  | quantidade:  70  | preco:  5  | avaliacao:  4
posicao:  2  | nome:  Livro Disney  | codigo:  497  | quantidade:  41  | preco:  2  | avaliacao:  4
posicao:  3  | nome:  Lapis Big  | codigo:  268  | quantidade:  76  | preco:  9  | avaliacao:  3
posicao:  4  | nome:  Lapis Foroni  | codigo:  29  | quantidade:  17  | preco:  8  | avaliacao:  4
posicao:  5  | nome:  Livro Disney  | codigo:  616  | quantidade:  7  | preco:  4  | avaliacao:  5
posicao:  6  | nome:  Caneta Marvel  | codigo:  276  | quantidade:  80  | preco:  9  | avaliacao:  3
posicao:  7  | nome:  Lapis Disney  | codigo:  436  | quantidade:  94  | preco:  3  | avaliacao:  2
posicao:  8  | nome:  Caneta Disney  | codigo:  626  | quantidade:  42  | preco:  3  | avaliacao:  4
posicao:  9  | nome:  Livro Foroni  | codigo:  681  | quantidade:  96  | preco:  3  | avaliaca

*Inserir 1000 registros*

In [486]:
elementos = []

for i in range(1000):
  nomes = ['Livro', 'Caderno', 'Caneta', 'Borracha', 'Lapis']
  marcas = ['Big', 'Fabber Castel', 'Foroni', 'Disney', 'Marvel']
  nome_produto = random.choice(nomes) + " " + random.choice(marcas)
  novo_produto = Produto(
    nome_produto,
    str(random.randint(1, 1000)),
    str(random.randint(1, 100)),
    str(random.randint(1, 10)),
    str(random.randint(1, 5)))
  elementos.append(novo_produto)

In [487]:
def insere_produto(elementos):
  for i in elementos:
    estocar.adicionar(i)
  return agendar


In [ ]:
%timeit insere_produto(elementos)

*Buscar por 1000 registros*

In [474]:
def busca_produto(elementos):
  for i in elementos:
    estocar.buscar(i.codigo, 'codigo')
  return agendar

In [475]:
%timeit busca_produto(elementos)

117 ms ± 13.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


*Ordenar*

In [221]:
estocar.ordenar('avaliacao')

indice 0 - Nome: Caneta Disney - código: 722, quantidade: 96, preço: 9, avaliações: 5
indice 1 - Nome: Caneta Disney - código: 297, quantidade: 77, preço: 4, avaliações: 5
indice 2 - Nome: Livro Fabber Castel - código: 859, quantidade: 39, preço: 6, avaliações: 5
indice 3 - Nome: Lapis Marvel - código: 540, quantidade: 100, preço: 10, avaliações: 5
indice 4 - Nome: Caneta Disney - código: 471, quantidade: 96, preço: 10, avaliações: 4
indice 5 - Nome: Livro Disney - código: 363, quantidade: 34, preço: 7, avaliações: 3
indice 6 - Nome: Lapis Marvel - código: 197, quantidade: 9, preço: 7, avaliações: 3
indice 7 - Nome: Livro Big - código: 730, quantidade: 64, preço: 6, avaliações: 3
indice 8 - Nome: Livro Marvel - código: 44, quantidade: 38, preço: 4, avaliações: 2
indice 9 - Nome: Borracha Marvel - código: 819, quantidade: 84, preço: 10, avaliações: 1


In [490]:
estocar.calcular()

292

*Buscar*

In [377]:
estocar.buscar("Livro Marvel", 'nome')

9

*Excluir*

In [363]:
estocar.excluir(212)

In [364]:
estocar.visualizar()

posicao:  0  | nome:  Livro Foroni  | codigo:  876  | quantidade:  44  | preco:  8  | avaliacao:  2
posicao:  1  | nome:  Livro Foroni  | codigo:  860  | quantidade:  41  | preco:  7  | avaliacao:  5
posicao:  2  | nome:  Lapis Disney  | codigo:  704  | quantidade:  18  | preco:  3  | avaliacao:  4
posicao:  3  | nome:  Caderno Big  | codigo:  970  | quantidade:  30  | preco:  10  | avaliacao:  4
posicao:  4  | nome:  Livro Foroni  | codigo:  478  | quantidade:  38  | preco:  9  | avaliacao:  4
posicao:  5  | nome:  Livro Foroni  | codigo:  486  | quantidade:  26  | preco:  7  | avaliacao:  2
posicao:  6  | nome:  Lapis Disney  | codigo:  521  | quantidade:  99  | preco:  5  | avaliacao:  1
posicao:  7  | nome:  Caderno Foroni  | codigo:  270  | quantidade:  3  | preco:  4  | avaliacao:  1
posicao:  8  | nome:  Livro Disney  | codigo:  865  | quantidade:  2  | preco:  1  | avaliacao:  4


4) Responda:

**a) Qual foi o melhor tempo de execução de adição e de qual sistema?**
Inserção no sistema de Agenda, com 312 ms


**b) Qual foi o melhor tempo de execução de busca e de qual sistema?**
Busca Binária no sistema de Agenda, com 6.95 ms


Agenda inserção: 315 ms O(n)

Agenda Busca: 6.95 ms

Produto inserção: 319 ms O(1)

Produto Busca: 117 ms